modules/core-components/src/test/java/org/mule/runtime/core/internal/source/polling/DefaultSchedulerMessageSourceTestCase.java

```diff
@@ -126,6 +130,63 @@ public String describeFailure() {
     });
   }
 
+  @Test
+  @Issue("W-17981248")
+  public void testStartWhenStopping() throws Exception {
+    DefaultSchedulerMessageSource schedulerMessageSource = createMessageSource();
+    SensingNullMessageProcessor flow = getSensingNullMessageProcessor();
+    schedulerMessageSource.setListener(flow);
+    schedulerMessageSource.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+
+    DefaultMuleContext spyMuleContext = spy((DefaultMuleContext) muleContext);
+    doReturn(true).when(spyMuleContext).isStopping();
+
+    schedulerMessageSource.setMuleContext(spyMuleContext);
+    schedulerMessageSource.start();
+
+    assertThat("Scheduler should not be started when MuleContext is stopping",
+               schedulerMessageSource.isStarted(), is(false));
+  }
+
+  @Test
+  @Issue("W-17981248")
+  public void testRunWhenStopping() throws Exception {
+    DefaultSchedulerMessageSource schedulerMessageSource = createMessageSource();
+    SensingNullMessageProcessor flow = getSensingNullMessageProcessor();
+    schedulerMessageSource.setListener(flow);
+    schedulerMessageSource.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+
+    DefaultMuleContext spyMuleContext = spy((DefaultMuleContext) muleContext);
+    doReturn(false).when(spyMuleContext).isStopping();
+    doReturn(true).when(spyMuleContext).isPrimaryPollingInstance();
+
+    schedulerMessageSource.setMuleContext(spyMuleContext);
+    schedulerMessageSource.start();
+    doReturn(true).when(spyMuleContext).isStopping();
+
+    verify(spyMuleContext, atLeastOnce()).isStopping();
+    // isPrimaryPollingInstance() was never called if isStopping() returns true
```

Comment:
The below line is useful for the test because isPrimaryPollingInstance() is only invoked when the source already checked if isStopping(), it returned true and it will execute the task (what you want to assert is that the task is not executed)